cache/store_test.go

```diff
@@ -0,0 +1,445 @@
+			initialRev:  6,
+			start:       []byte("/z"),
+			end:         []byte("/a"),
+			expectedKVs: nil,
+			expectedRev: 6,
+		},
+		{
+			name:        "Get_fromKey_scans_ordered",
+			initialKVs:  []*mvccpb.KeyValue{makeKV("/a", "1", 7), makeKV("/b", "2", 7), makeKV("/c", "3", 7), makeKV("/d", "4", 7)},
+			initialRev:  7,
+			start:       []byte("/b"),
+			end:         []byte{0},
+			expectedKVs: []*mvccpb.KeyValue{makeKV("/b", "2", 7), makeKV("/c", "3", 7), makeKV("/d", "4", 7)},
+			expectedRev: 7,
+		},
+		{
+			name:        "Get_fromKey_with_no_results",
+			initialKVs:  []*mvccpb.KeyValue{makeKV("/a", "1", 9), makeKV("/b", "2", 9)},
+			initialRev:  9,
+			start:       []byte("/zzz"),
+			end:         []byte{0},
+			expectedKVs: nil,
+			expectedRev: 9,
+		},
+	}
+
+	for _, tt := range tests {
+		test := tt
+		t.Run(test.name, func(t *testing.T) {
+			s := newStore(8)
+			if test.initialKVs != nil {
+				s.Restore(test.initialKVs, test.initialRev)
+			}
+
+			kvs, rev, err := s.Get(test.start, test.end)
+
+			if test.expectedErr != nil {
+				if !errors.Is(err, test.expectedErr) {
+					t.Fatalf("Get error = %v; want %v", err, test.expectedErr)
+				}
+				return
+			}
+			if err != nil {
+				t.Fatalf("Get returned unexpected error: %v", err)
+			}
+			if rev != test.expectedRev {
+				t.Fatalf("revision=%d; want %d", rev, test.expectedRev)
+			}
+
+			if diff := cmp.Diff(test.expectedKVs, kvs); diff != "" {
+				t.Fatalf("Get mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestStoreApply(t *testing.T) {
+	type testCase struct {
+		name         string
+		initialKVs   []*mvccpb.KeyValue
+		initialRev   int64
+		eventBatches [][]*clientv3.Event
+
+		expectedLatestRev int64
+		expectedSnapshot  []*mvccpb.KeyValue
+		expectErr         bool
+	}
+	tests := []testCase{
+		{
+			name:              "apply_with_same_key",
+			initialKVs:        []*mvccpb.KeyValue{makeKV("/k", "v1", 10)},
+			initialRev:        10,
+			eventBatches:      [][]*clientv3.Event{{makePutEvent("/k", "v2", 11)}},
+			expectedLatestRev: 11,
+			expectedSnapshot:  []*mvccpb.KeyValue{makeKV("/k", "v2", 11)},
+		},
+		{
+			name:       "apply_mixed_operations",
+			initialKVs: []*mvccpb.KeyValue{makeKV("/a", "oldA", 10), makeKV("/b", "oldB", 10)},
+			initialRev: 10,
+			eventBatches: [][]*clientv3.Event{
+				{makePutEvent("/a", "newA", 11)},
+				{makeDelEvent("/b", 12), makePutEvent("/c", "C", 12)},
+			},
+			expectedLatestRev: 12,
+			expectedSnapshot:  []*mvccpb.KeyValue{makeKV("/a", "newA", 11), makeKV("/c", "C", 12)},
+		},
+		{
+			name:       "apply_with_same_revision_multiple_events",
+			initialKVs: []*mvccpb.KeyValue{makeKV("/a", "oldA", 10), makeKV("/b", "oldB", 10)},
+			initialRev: 10,
+			eventBatches: [][]*clientv3.Event{
+				{makePutEvent("/a", "newA", 10), makePutEvent("/b", "newB", 10)},
+			},
+			expectedLatestRev: 10,
+			expectedSnapshot:  []*mvccpb.KeyValue{makeKV("/a", "newA", 10), makeKV("/b", "newB", 10)},
+		},
```

Comment:
Note, events with single revision should be applied at once, within single lock, same like etcd. If not, cache will expose partially applied revision which is invalid. So this test case is incorrect, also please make sure we correctly apply revision. We could add validation in storage that prevents same revision being applied twice to prevent this kind of bug.